### PR TITLE
Improve lint for type alias bounds

### PR DESCRIPTION
**Pull request by RalfJung**
First of all, I learned just today that I was wrong assuming that the bounds in type aliases are entirely ignored: It turns out they are used to resolve associated types in type aliases. So:
```rust
type T1<U: Bound> = U::Assoc; // compiles
type T2<U> = U::Assoc; // fails
type T3<U> = <U as Bound>::Assoc; // "correct" way to write this, maybe?
```
I am sorry for creating this mess.

This PR changes the wording of the lint accordingly. Moreover, since just removing the bound is no longer always a possible fix, I tried to detect cases like `T1` above and show a helpful message to the user:
```
warning: bounds on generic parameters are not enforced in type aliases
  --> $DIR/type-alias-bounds.rs:57:12
   |
LL | type T1<U: Bound> = U::Assoc; //~ WARN not enforced in type aliases
   |            ^^^^^
   |
   = help: the bound will not be checked when the type alias is used, and should be removed
help: use absolute paths (i.e., <T as Trait>::Assoc) to refer to associated types in type aliases
  --> $DIR/type-alias-bounds.rs:57:21
   |
LL | type T1<U: Bound> = U::Assoc; //~ WARN not enforced in type aliases
   |                     ^^^^^^^^
```
I am not sure if I got this entirely right. Ideally, we could provide a suggestion involving the correct trait and type name -- however, while I have access to the HIR in the lint, I do not know how to get access to the resolved name information, like which trait `Assoc` belongs to above. The lint does not even run if that resolution fails, so I assume that information is available *somewhere*...

This is a follow-up for (parts of) https://github.com/rust-lang/rust/pull/48326. Also see https://github.com/rust-lang/rust/issues/21903.

This changes the name of a lint, but that lint was just merged to master yesterday and has never even been on beta.

### CODE REVIEW

**Comment by RalfJung** [comment]
I guess there is also the question if we even want to nudge people towards writing `type T3<U> = <U as Bound>::Assoc;`. If that is controversial, I suggest I remove the last commit from this MR (so we get the lint wording and name corrected at least).

The alternative to having people write `type T3<U> = <U as Bound>::Assoc;` is to detect if a bound is necessary for well-formedness. So, `type T3<U: Bound> = U::Assoc;` should be accepted but `type T3<U: Bound> = Vec<U>;` should be linted against. I have no idea how to do that.

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -1316,48 +1317,115 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnreachablePub {
...
+            }
+            hir::QPath::Resolved(..) => false,
+        }
+    }
+
+    fn suggest_changing_assoc_types(ty: &hir::Ty, err: &mut DiagnosticBuilder) {
+        // Access to associates types should use `<T as Bound>::Assoc`, which does not need a
+        // bound.  Let's see of this type does that.
+
+        // We use an AST visitor to walk the type.
```
Comment (petrochenkov) [inline src/librustc_lint/builtin.rs]: Nit: HIR visitor

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -1316,48 +1317,115 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnreachablePub {
...
+        impl<'a, 'db, 'v> Visitor<'v> for WalkAssocTypes<'a, 'db> {
+            fn nested_visit_map<'this>(&'this mut self) -> intravisit::NestedVisitorMap<'this, 'v>
+            {
+                intravisit::NestedVisitorMap::None
+            }
+
+            fn visit_qpath(&mut self, qpath: &'v hir::QPath, id: NodeId, span: Span) {
+                if TypeAliasBounds::is_type_variable_assoc(qpath) {
+                    self.err.span_help(span,
+                        "use absolute paths (i.e., <T as Trait>::Assoc) to refer to associated \
```
Comment (petrochenkov) [inline src/librustc_lint/builtin.rs]: Absolute paths are `::a::b::c`.
IIRC, doc and lang teams had a council at one time and discussed how to call "UFCS paths" (`<T as Trait>::Assoc`) properly, but I don't remember what they decided. "Fully disambiguated paths" maybe.

Comment (petrochenkov) [reply src/librustc_lint/builtin.rs]: Also, missing backquotes around `<T as Trait>::Assoc`.

**Comment by petrochenkov** [comment]
r=me modulo couple of wording nits
@bors delegate+

**Comment by petrochenkov** [comment]
\+ Travis says some compile-fail tests are failing

>[01:04:45]     [compile-fail] compile-fail/issue-17994.rs
>[01:04:45]     [compile-fail] compile-fail/private-in-public-warn.rs

**Comment by RalfJung** [comment]
> + Travis says some compile-fail tests are failing

It seems those warnings went away when I moved this from an early to a late lint... is it possible that late (HIR) lints don't run in these cases?

**Comment by RalfJung** [comment]
> is it possible that late (HIR) lints don't run in these cases?

@rkruppe confirmed on IRC that's a reasonable assumption, and the UI test looking for these warnings still passes so I think all is good.

@bors r=@petrochenkov

**Pull request merged**